syncer/client.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return nil, err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), defaultTimeoutForStatus)
+	defer cancel()
+
+	status, err := clt.GetStatus(timeoutCtx, &emptypb.Empty{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &NoForkPeer{
+		ID:       peerID,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(peerID),
+	}, nil
+}
+
+func (m *syncPeerClient) GetConnectedPeerStatuses() []*NoForkPeer {
+	var (
+		ps            = m.network.Peers()
+		syncPeers     = make([]*NoForkPeer, 0, len(ps))
+		syncPeersLock = sync.Mutex{}
+		wg            sync.WaitGroup
+	)
+
+	for _, p := range ps {
+		p := p
+
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			peerID := p.Info.ID
+
+			status, err := m.GetPeerStatus(peerID)
+			if err != nil {
+				m.logger.Warn("failed to get status from a peer, skip", "id", peerID, "err", err)
+			}
+
+			syncPeersLock.Lock()
+
+			syncPeers = append(syncPeers, status)
+
+			syncPeersLock.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	return syncPeers
+}
+
+func (m *syncPeerClient) GetPeerStatusUpdateCh() <-chan *NoForkPeer {
+	return m.peerStatusUpdateCh
+}
+
+func (m *syncPeerClient) GetPeerConnectionUpdateEventCh() <-chan *event.PeerEvent {
+	return m.peerConnectionUpdateCh
+}
+
+func (m *syncPeerClient) startGossip() error {
+	topic, err := m.network.NewTopic(statusTopicName, &proto.SyncPeerStatus{})
+	if err != nil {
+		return err
+	}
+
+	if err := topic.Subscribe(m.handleStatusUpdate); err != nil {
+		return fmt.Errorf("unable to subscribe to gossip topic, %w", err)
+	}
+
+	m.topic = topic
+
+	return nil
+}
+
+func (m *syncPeerClient) handleStatusUpdate(obj interface{}, from peer.ID) {
+	status, ok := obj.(*proto.SyncPeerStatus)
+	if !ok {
+		m.logger.Error("failed to cast gossiped message to txn")
+
+		return
+	}
+
+	if !m.network.IsConnected(from) {
+		if m.id != from.String() {
+			m.logger.Warn("received status from non-connected peer, ignore", "id", from)
+		}
+
+		return
+	}
+
+	m.peerStatusUpdateCh <- &NoForkPeer{
+		ID:       from,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(from),
+	}
+}
+
+func (m *syncPeerClient) startNewBlockProcess() {
+	m.subscription = m.blockchain.SubscribeEvents()
+
+	for event := range m.subscription.GetEventCh() {
+		if event.Type == blockchain.EventHead && len(event.NewChain) > 0 {
+			latest := m.blockchain.Header()
+			if latest == nil {
+				return
+			}
+
+			// Publish status
+			if err := m.topic.Publish(&proto.SyncPeerStatus{
+				Number: latest.Number,
+			}); err != nil {
+				m.logger.Warn("failed to publish status", "err", err)
+			}
+		}
+	}
+}
+
+func (m *syncPeerClient) startPeerEventProcess() {
+	peerEventCh, err := m.network.SubscribeCh()
+	if err != nil {
+		m.logger.Error("failed to subscribe", "err", err)
+
+		return
+	}
+
+	for e := range peerEventCh {
+		switch e.Type {
+		case event.PeerConnected, event.PeerDisconnected:
+			m.peerConnectionUpdateCh <- e
+		}
```

Comment:
Do you even need a switch here?
This can be simplified

Reply:
Fixed for you at d75dc06974a15bbf55ba2bc69f8173b95e531db2